nixos/modules/services/tracing/tempo.nix

```diff
@@ -1,60 +1,546 @@
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Whether to enable access key ENV VAR usage for static credentials.
+
+        If enabled, a file is expected to exist at location specified by option
+        storageS3AccessCredsPath.
+      '';
+    };
+
+    storageS3AccessCredsPath = mkOption {
+      type = str;
+      default = "/run/keys/tempo";
+      description = mdDoc ''
+        Specifies the location of an S3 credentials file that Tempo should utilize.
+
+        This file is expected to contain the following substituted lines:
+
+        AWS_ACCESS_KEY_ID=$SECRET_KEY_ID
+        AWS_SECRET_ACCESS_KEY=$SECRET_KEY
+      '';
+    };
+
+    storageS3ForcePathStyle = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Enable to use path-style requests.
+      '';
+    };
+
+    storageS3Insecure = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Debugging option for temporary http testing.
+      '';
+    };
+
+    storageS3InsecureSkipVerify = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Debugging option for temporary https testing.
+      '';
+    };
+
+    storageTraceWalPath = mkOption {
+      type = str;
+      default = "/var/lib/tempo/storage/wal";
+      description = mdDoc ''
+        Where to store the head blocks while they are being appended to.
+      '';
+    };
+
+    searchEnable = mkOption {
+      type = bool;
+      default = true;
+      description = mdDoc ''
+        Enable tempo search.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = attrs;
+      default = {};
+      description = mdDoc ''
+        Extra configuration to pass to Tempo service.
         See https://grafana.com/docs/tempo/latest/configuration/ for available options.
       '';
     };
 
     configFile = mkOption {
-      type = types.nullOr types.path;
+      type = nullOr path;
       default = null;
-      description = lib.mdDoc ''
-        Specify a path to a configuration file that Tempo should use.
+      description = mdDoc ''
+        As an alternative to building a working Tempo configuration using the available
+        module options, the path to a complete Tempo configuration file may be provided.
+
+        When a path to a complete Tempo configuration file is provided, all Tempo module
+        options will be ignored in favor of the Tempo configuration file, with the
+        exception of the 'service.tempo.enable' and 'services.tempo.logLevel' options.
       '';
     };
   };
 
   config = mkIf cfg.enable {
-    # for tempo-cli and friends
-    environment.systemPackages = [ pkgs.tempo ];
+    assertions = mkIf (cfg.configFile == null) [
+      {
+        assertion =
+          if cfg.metricsGeneratorEnableServiceGraphs || cfg.metricsGeneratorEnableSpanMetrics
+          then cfg.metricsGeneratorStorageRemoteWrite != null
+          else true;
```

Comment:
Please write this in one boolean expression without if then else. Same for the other two.

Reply:
Done with logical implication